src/vr/clock.zig

```diff
@@ -0,0 +1,508 @@
+                self.marzullo_tuples[count] = Marzullo.Tuple{
+                    .source = @intCast(u8, source),
+                    .offset = sample.clock_offset - @intCast(i64, sample.one_way_delay + tolerance),
+                    .bound = .lower,
+                };
+                count += 1;
+                self.marzullo_tuples[count] = Marzullo.Tuple{
+                    .source = @intCast(u8, source),
+                    .offset = sample.clock_offset + @intCast(i64, sample.one_way_delay + tolerance),
+                    .bound = .upper,
+                };
+                count += 1;
+            }
+        }
+        return self.marzullo_tuples[0..count];
+    }
+
+    fn minimum_one_way_delay(a: ?Sample, b: ?Sample) ?Sample {
+        if (a == null) return b;
+        if (b == null) return a;
+        if (a.?.one_way_delay < b.?.one_way_delay) return a;
+        // Choose B if B's one way delay is less or the same (we assume B is the newer sample):
+        return b;
+    }
+};
+
+pub const SystemTime = struct {
+    const Self = @This();
+
+    /// Hardware and/or software bugs can mean that the monotonic clock may regress.
+    /// One example (of many): https://bugzilla.redhat.com/show_bug.cgi?id=448449
+    /// We crash the process for safety if this ever happens, to protect against infinite loops.
+    /// It's better to crash and come back with a valid monotonic clock than get stuck forever.
+    monotonic_guard: u64 = 0,
+
+    /// A timestamp to measure elapsed time, meaningful only on the same system, not across reboots.
+    /// Always use a monotonic timestamp if the goal is to measure elapsed time.
+    /// This clock is not affected by discontinuous jumps in the system time, for example if the
+    /// system administrator manually changes the clock.
+    pub fn monotonic(self: *Self) u64 {
+        // The true monotonic clock on Linux is not in fact CLOCK_MONOTONIC:
+        // CLOCK_MONOTONIC excludes elapsed time while the system is suspended (e.g. VM migration).
+        // CLOCK_BOOTTIME is the same as CLOCK_MONOTONIC but includes elapsed time during a suspend.
+        // For more detail and why CLOCK_MONOTONIC_RAW is even worse than CLOCK_MONOTONIC,
+        // see https://github.com/ziglang/zig/pull/933#discussion_r656021295.
+        var ts: std.os.timespec = undefined;
+        std.os.clock_gettime(std.os.CLOCK_BOOTTIME, &ts) catch unreachable;
+        const m = @intCast(u64, ts.tv_sec) * std.time.ns_per_s + @intCast(u64, ts.tv_nsec);
+        assert(m >= self.monotonic_guard);
+        self.monotonic_guard = m;
+        return m;
+    }
+
+    /// A timestamp to measure real (i.e. wall clock) time, meaningful across systems, and reboots.
+    /// This clock is affected by discontinuous jumps in the system time.
+    pub fn realtime(self: *Self) i64 {
+        var ts: std.os.timespec = undefined;
+        std.os.clock_gettime(std.os.CLOCK_REALTIME, &ts) catch unreachable;
```

Comment:
`unreachable` is correct here according to the `clock_gettime(2)` man page as "All implementations support the system-wide real-time clock, which is identified by CLOCK_REALTIME."